micromanager_gui/main_window.py

```diff
@@ -148,92 +159,121 @@ def add_frame_explorer(self, name, array):
         except KeyError:
             self.viewer.add_image(array, name=layer_name)
 
-    # TO DO: add the file name form the save box
-    def _on_mda_frame(self, image, event):
 
-        sequence = event.sequence
+    # create temp folder when mda starts
+    def _on_mda_started_temp_folder(self, sequence: useq.MDASequence):
 
-        event.index.setdefault('t', 0)
-        event.index.setdefault('z', 0)
+        self.temp_folder = tempfile.TemporaryDirectory(None, str(sequence.uid))
+        return self.temp_folder.name
 
-        t_stack_length = len(sequence.time_plan) or 1
-        p_stack_length = len(sequence.stage_positions)
-        z_stack_length = len(sequence.z_plan) or 1
-        c_stack_length = len(sequence.channels)
+    #get the name to use to save the active ayer
+    def get_filename(self, fname, list_dir):
+        val = int(fname.split('_')[-1])
+        while True:
+            new_val = '{0:03}'.format(val)
+            fname = fname[:-3] + new_val
+            if ((fname + '.tif') or (fname + '.tiff')) not in set(list_dir):
+                break
+            else:
+                val += 1
+        return fname
 
-        event_index_t = event.index["t"]
-        event_index_p = event.index["p"]
-        event_index_z = event.index["z"]
-        event_index_c = event.index["c"]
-        
-        file_name = 'mda'
+    # delete temp folder and files when mda is done and save layer
+    # and add increment to save name
+    def _on_mda_finished_save_layer(self, sequence: useq.MDASequence):
+            
+        if self.mda.save_groupBox.isChecked():
 
-        layer_name = (
-            f'{file_name}_[{sequence.axis_order}]_p{p_stack_length}_'
-            f't{t_stack_length}_z{z_stack_length}_c{c_stack_length}'
-        )
+            active_layer = self.viewer.layers.selection.active.name
```

Comment:
this makes the assumption that the user hasn't changed the layer in the meantime

Reply:
I guess this might work:
```
layer_uid = [l.name for l in self.viewer.layers if l.metadata['uid'] == sequence.uid]
active_layer = self.viewer.layers[str(layer_uid[0])]
```

Reply:
you're already looping through `viewer.layers` ... why not just grab the layer directly rather than just taking the name so that you then index into viewer.layers again...:

```
for layer in self.viewer.layers:
    if layer.metadata.get('uid') == sequence.uid:
        sequence_layer = layer
        break
else:
    raise IndexError("could not find layer corresponding to sequence")
```

note also, it's very important to use `if l.metadata.get('uid')` instead of `if l.metadata['uid']`, since the former will raise an exception if it encounters a layer that doesn't have the `uid` key in the metadata dict (which will basically be every other layer)

Reply:
alternate syntax:
```python
try:
    active_layer = next(l for l in self.viewer.layers if l.metadata.get('uid') == sequence.uid)
except StopIteration:
    raise IndexError("could not find layer corresponding to sequence")
```